make module! arguments optional

kloenk opened the pull request:
Fixed: #215

wedsonaf commented:
Can you move the parsing of `ModuleInfo` into a function so that we can use it in `module_misc_device` as well?

kloenk commented:
> Can you move the parsing of `ModuleInfo` into a function so that we can use it in `module_misc_device` as well?

Done, also added alias to the `module_misc_device` and refactored it to use the new `ModuleInfo` struct

ojeda commented on rust/module.rs:
```diff
@@ -103,6 +103,7 @@ fn expect_end(it: &mut token_stream::IntoIter) {
     }
 }
 
+#[allow(dead_code)]
```
Remove the dead code, if we need them we can reintroduce them later.

ojeda commented on rust/module.rs:
```diff
@@ -315,6 +325,102 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+impl ModuleInfo {
+    fn parse(it: &mut token_stream::IntoIter) -> Self {
+        let mut info = ModuleInfo::default();
+        // the first allowed position is 1, to prevent setting something twice
```
Nit:

```suggestion
        // The first allowed position is 1, to prevent setting something twice.
```

ojeda commented on samples/rust/rust_chrdev.rs:
```diff
@@ -16,8 +16,7 @@ module! {
     author: b"Rust for Linux Contributors",
     description: b"Rust character device sample",
     license: b"GPL v2",
-    params: {
-    },
```
Please remove the empty `params` from the other samples and from binder, since now we can :)

If you move the removals to another commit, that would be ideal.

ojeda commented on samples/rust/rust_chrdev.rs:
```diff
@@ -16,8 +16,7 @@ module! {
     license: b"GPL v2",
-    params: {
-    },
+    alias: b"samples_rust_chrdev",
```
Instead of adding an alias to this sample, I would improve the docs of the `module!` macro to include an explanation/example of all the optional keys etc.

kloenk replied:
Added it to the documentation, could you please check for spelling mistakes :-)

ojeda commented on rust/module.rs:
```diff
@@ -315,6 +307,102 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+                let params = expect_group(it);
+                info.params = Some(params);
+            } else {
+
```
```suggestion
```

ojeda commented on rust/module.rs:
```diff
@@ -365,6 +453,16 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+///   - `license`: byte array of the license of the kernel module (required).
+///   - `alias`: byte array of alias name of the kernel module.
+///   - `alias_rtnl_link`: byte array of the rtnl_link_alias of the kernel module (mutually exclusive with `alias`).
+///   - `params`: parameters for the kernel module, as decsribed below.
```
```suggestion
///   - `params`: parameters for the kernel module, as described below.
```

ojeda commented on rust/module.rs:
```diff
@@ -365,6 +453,16 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+///   - `description`: byte array of the description of the kernel module.
+///   - `license`: byte array of the license of the kernel module (required).
+///   - `alias`: byte array of alias name of the kernel module.
+///   - `alias_rtnl_link`: byte array of the rtnl_link_alias of the kernel module (mutually exclusive with `alias`).
```
```suggestion
///   - `alias_rtnl_link`: byte array of the `rtnl_link_alias` of the kernel module (mutually exclusive with `alias`).
```

ojeda commented on rust/module.rs:
```diff
@@ -315,6 +307,102 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+
+        expect_end(it);
+
+        info.check_required();
```
Nit: I would paste the definition here and remove the method, since we only call it once, but both ways are fine.

ojeda commented on rust/module.rs:
```diff
@@ -315,6 +307,102 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+    }
+}
+
+fn allowed_pos(pos: &mut u8, allowed_pos: u8, name: &str) {
+    if *pos > allowed_pos {
+        panic!("'{}' is not allowed at this position", name);
+    }
+    *pos = allowed_pos;
+}
```
This one could be inside the `impl`.

Or better, a local lambda so that we only need to write `allowed_pos(x)`.

kloenk replied:
I can't get my head wrapped around making it a lambda at the moment. will put it inside the `impl`.

ojeda commented on rust/module.rs:
```diff
@@ -746,18 +837,19 @@ pub fn module_misc_device(ts: TokenStream) -> TokenStream {
-        license = license
+        type_ = info.type_,
+        name = info.name,
+        author = info.author.map(|v| format!("author: b\"{}\",", v)).unwrap_or_else(|| "".to_string()),
+        description = info.description.map(|v| format!("description: b\"{}\",", v)).unwrap_or_else(|| "".to_string()),
+        alias = info.alias.map(|v| format!("alias: b\"{}\",", v)).unwrap_or_else(|| "".to_string()),
```
Perhaps we could factor this out, in particular when we get new specialized `module_*!` macros, but it is not a big deal.

ojeda commented on rust/module.rs:
```diff
@@ -315,6 +307,97 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+
+        info
+    }
+    fn allowed_pos(pos: &mut u8, allowed_pos: u8, name: &str) {
```
Newline between methods/functions please.

kloenk replied:
I'm sorry, just copied it around. Also rustfmt(check) does not work with O=build. did not see that until now.

ojeda replied:
> Also rustfmt(check) does not work with O=build. did not see that until now.

Fixed on #240.

ojeda commented on rust/module.rs:
```diff
@@ -388,178 +481,174 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+        assert_eq!(params.delimiter(), Delimiter::Brace);
+
+        let mut it = params.stream().into_iter();
+
+
```
Double empty line.

ojeda commented on rust/module.rs:
```diff
@@ -315,6 +307,97 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
+impl ModuleInfo {
+    fn parse(it: &mut token_stream::IntoIter) -> Self {
+        let mut info = ModuleInfo::default();
+        // The first allowed position is 1, to prevent setting something twice
```
We are ending the comments with a period (I am documenting this).

ojeda reviewed (changes requested):
Since you have to fix the `rustfmt` formatting anyway, I took the change to add a couple more comments...
